File: android/app/src/main/java/com/example/android/viewmodel/home/HomeViewModel.kt
package com.example.android.viewmodel.home

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.android.data.model.RecipeRequestPayload
import com.example.android.data.model.RecipeSuggestion
import com.example.android.data.repository.ApiRecipeRepository
import com.example.android.data.repository.RecipeRepository
import com.example.android.viewmodel.route.OptimizedRouteStop
import com.example.android.viewmodel.route.RoutePlan
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

data class HomeMetric(
    val label: String,
    val value: String,
    val sublabel: String
)

data class RouteStop(
    val index: Int,
    val name: String,
    val address: String,
    val distance: String,
    val timeEstimate: String,
    val itemsLabel: String
)

data class HomeUiState(
    val userName: String = "John Doe",
    val savingsThisTrip: String = "$6.70",
    val totalBudget: String = "$120.00",
    val budgetUsed: String = "$57.31",
    val savedThisMonth: String = "$42.80",
    val savedThisMonthLabel: String = "this month",
    val avgTripTime: String = "34m",
    val milesSaved: String = "12.4",
    val itemsTracked: String = "89",
    val itemsTrackedLabel: String = "tracked",
    val alertsCount: String = "3",
    val optimizedStopsLabel: String = "3 stops",
    val metrics: List<HomeMetric> = emptyList(),
    val featuredRecipe: RecipeSuggestion? = null,
    val isLoadingRecipe: Boolean = false,
    val recipeError: String? = null
)

/**
 * Demo route preview shown on Home when no real optimized route is available yet.
 * Mirrors the original hardcoded sample so the screen still has visual content for
 * a fresh user. Once `RouteViewModel.uiState.optimizedRoute` is non-null, the
 * real route is rendered instead — see [HomeScreen].
 */
val DemoRouteStops: List<RouteStop> = listOf(
    RouteStop(1, "Aldi", "142 Atlantic Ave", "0.8 mi", "12 min", "3 items"),
    RouteStop(2, "Trader Joe's", "130 Court St", "1.2 mi", "8 min", "4 items"),
    RouteStop(3, "Costco", "976 3rd Ave", "2.4 mi", "15 min", "2 items")
)

/**
 * Map up to the first three stops of a real optimized route into the lightweight
 * [RouteStop] preview shape used on Home. Distance / time are not yet available
 * on `OptimizedRouteStop` (Mapbox wires that up in S1.5), so we leave them blank.
 * Items label uses the per-stop quantity sum.
 */
fun previewStopsFrom(activeRoute: RoutePlan): List<RouteStop> {
    return activeRoute.stops.take(PREVIEW_STOP_LIMIT).mapIndexed { index, stop ->
        RouteStop(
            index = index + 1,
            name = stop.storeName,
            address = stop.address ?: "Address not yet available",
            distance = "",
            timeEstimate = "",
            itemsLabel = stop.itemsLabel()
        )
    }
}

private const val PREVIEW_STOP_LIMIT = 3

private fun OptimizedRouteStop.itemsLabel(): String {
    val total = items.sumOf { it.quantity }
    return when (total) {
        0 -> "No items"
        1 -> "1 item"
        else -> "$total items"
    }
}

class HomeViewModel(
    private val recipeRepository: RecipeRepository = ApiRecipeRepository()
) : ViewModel() {
    var uiState by mutableStateOf(
        HomeUiState(
            metrics = listOf(
                HomeMetric("Budget", "$57.31", "of $120.00"),
                HomeMetric("Saved", "$42.80", "this month"),
                HomeMetric("Avg trip", "34m", ""),
                HomeMetric("Miles saved", "12.4", "miles saved"),
                HomeMetric("Items", "89", "tracked"),
                HomeMetric("Alerts", "3", "alerts")
            )
        )
    )
        private set

    private var lastPayload: RecipeRequestPayload? = null
    private var recipeJob: Job? = null

    /**
     * Generate (or refresh) the featured recipe. Mirrors iOS
     * `HomeViewModel.loadRecipe` (HomeViewModel.swift:56-74): if the payload
     * matches the last one served and we already have a recipe, skip the call
     * to avoid a loading flicker. The repository also caches by payload, so a
     * `force = true` always re-requests.
     */
    fun refreshRecipe(payload: RecipeRequestPayload, force: Boolean = false) {
        if (!force && payload == lastPayload && uiState.featuredRecipe != null) {
            return
        }
        recipeJob?.cancel()
        uiState = uiState.copy(isLoadingRecipe = true, recipeError = null)
        recipeJob = viewModelScope.launch {
            val result = recipeRepository.generate(payload, force)
            result.fold(
                onSuccess = { recipe ->
                    lastPayload = payload
                    uiState = uiState.copy(
                        featuredRecipe = recipe,
                        isLoadingRecipe = false,
                        recipeError = null
                    )
                },
                onFailure = { error ->
                    uiState = uiState.copy(
                        isLoadingRecipe = false,
                        recipeError = error.message ?: "Unable to generate a recipe right now."
                    )
                }
            )
        }
    }

    /** Force-regenerate the recipe regardless of cached payload. */
    fun retryRecipe(payload: RecipeRequestPayload) {
        refreshRecipe(payload, force = true)
    }
}

File: ios/Neighborly/Neighborly/ViewModels/HomeViewModel.swift
import Foundation
import SwiftUI
import CoreLocation

struct HomeMetric: Identifiable {
    let id = UUID()
    let label: String
    let value: String
    let sublabel: String
    let icon: String
    let tint: MetricTint

    enum MetricTint {
        case green, orange, gray
    }
}

struct HomeRouteStop: Identifiable {
    let id = UUID()
    let index: Int
    let name: String
    let address: String
    let distance: String
    let timeEstimate: String
    let itemsLabel: String
}

@Observable
final class HomeViewModel {
    var featuredRecipe: RecipeSuggestion?
    var isLoadingRecipe = false
    var recipeError: String?

    private var lastRecipeRequest: RecipeRequestPayload?
    /// owned by the model so the in-flight request survives view lifecycle
    /// (e.g. user switching tabs mid-load). without this, .task cancellation
    /// would discard the response and we'd have wasted backend credits.
    private var loadTask: Task<Void, Never>?

    let userName: String = "John Doe"
    let savingsThisTrip: String = "6.70"
    let totalBudget: String = "$120.00"
    let budgetUsed: String = "$57.31"
    let budgetProgress: Double = 57.31 / 120.0
    let savedThisMonth: String = "$42.80"
    let savedThisMonthLabel: String = "this month"
    let avgTripTime: String = "34m"
    let milesSaved: String = "12.4"
    let itemsTracked: String = "89"
    let alertsCount: String = "3"
    let optimizedStopsLabel: String = "3 stops"

    let savingsBarHeights: [CGFloat] = [0.4, 0.7, 0.5, 0.9, 0.6]

    let routeStops: [HomeRouteStop] = [
        HomeRouteStop(index: 1, name: "Aldi", address: "142 Atlantic Ave", distance: "0.8 mi", timeEstimate: "12 min", itemsLabel: "3 items"),
        HomeRouteStop(index: 2, name: "Trader Joe's", address: "130 Court St", distance: "1.2 mi", timeEstimate: "8 min", itemsLabel: "4 items"),
        HomeRouteStop(index: 3, name: "Costco", address: "976 3rd Ave", distance: "2.4 mi", timeEstimate: "15 min", itemsLabel: "2 items")
    ]

    /// Kick off a recipe load. Returns immediately — the actual work runs in
    /// a Task owned by this model, so it survives view disappearance.
    /// Subsequent calls with the same payload while one is already in-flight
    /// or already cached are no-ops.
    @MainActor
    func startLoadIfNeeded(
        using preferences: Preferences,
        userLocation: CLLocation? = nil,
        force: Bool = false
    ) {
        let payload = preferences.recipeRequestPayload(userLocation: userLocation)

        if !force {
            // already loading the same request → don't fire a duplicate
            if isLoadingRecipe, payload == lastRecipeRequest { return }
            // already have a cached result for these prefs → keep it
            if featuredRecipe != nil, payload == lastRecipeRequest { return }
        }

        // a force or a different payload supersedes any in-flight load
        loadTask?.cancel()

        isLoadingRecipe = true
        recipeError = nil
        lastRecipeRequest = payload

        loadTask = Task { [weak self] in
            do {
                let recipe = try await APIService.generateRecipe(
                    preferences: preferences,
                    userLocation: userLocation
                )
                if Task.isCancelled { return }
                await MainActor.run {
                    self?.featuredRecipe = recipe
                    self?.isLoadingRecipe = false
                }
            } catch is CancellationError {
                return
            } catch {
                if Task.isCancelled { return }
                await MainActor.run {
                    self?.featuredRecipe = nil
                    self?.recipeError = error.localizedDescription
                    self?.isLoadingRecipe = false
                }
            }
        }
    }
}
